syntax = "proto3";

option objc_class_prefix = "MMA";

package mma;

// define your service here

service MMA{
    rpc ReadPage (ReadRequest) returns (ReadReply) {}
    rpc WritePage (WriteRequest) returns (WriteReply) {}
    rpc Allocate (AllocateRequest) returns (AllocateReply) {}
    rpc Release (ReleaseRequest) returns (ReleaseReply) {}
}

message ReadRequest {
    int32 array_id = 1;
    int32 virtual_page_id = 2;
    int32 offset = 3;

}

message ReadReply {
    int32 data = 1;
    
}

message WriteRequest {
    int32 array_id = 1;
    int32 virtual_page_id = 2; 
    int32 offset = 3;
    int32 value = 4;
    
}

message WriteReply {
    
}

message AllocateRequest {
    int32 size = 1;
}

message AllocateReply {
    int32 array_id = 1;
}

message ReleaseRequest {
    int32 array_id = 1;
    int32 num_of_pages = 2;
}

message ReleaseReply {
    
}